mrmustard/lab/abstract/state.py

```diff
@@ -394,6 +345,117 @@ def primal(self, other: Union[State, Transformation]) -> State:
                 f"Cannot apply {other.__class__.__qualname__} to {self.__class__.__qualname__}"
             ) from e
 
+    def _project_onto_state(self, other: State) -> Union[State, float]:
+        """If states are gaussian use generaldyne measurement, else use
+        the states' Fock representation."""
+
+        # if both states are gaussian
+        if self.is_gaussian and other.is_gaussian:
+            return self._project_onto_gaussian(other)
+
+        # either self or other is not gaussian
+        return self._project_onto_fock(other)
+
+    def _project_onto_fock(self, other: State) -> Union[State, float]:
+        """Returns the post-measurement state of the projection between two non-Gaussian
+        states on the remaining modes or the probability of the result. When doing homodyne sampling,
+        returns the post-measurement state or the measument outcome if no modes remain.
+
+        Args:
+            other (State): state being projected onto self
+
+        Returns:
+            State or float: returns the output conditional state on the remaining modes
+                or the outcome of the homodyne measurement.
+        """
+
+        other_cutoffs = [
+            None if m not in self.modes else other.cutoffs[other.indices(m)] for m in other.modes
+        ]
+
+        projector_state = self  # state_m is the state used to project onto
+        outcome = None
+        # Only instances of `Homodyne` define the attribute "sampling". If this is
+        # True then homodyne sampling is performed.
+        if getattr(self, "sample", False):
+            # build pdf and sample homodyne outcome
+            outcome, projector_state = homodyne.sample_homodyne_fock(
+                state=other, quadrature_angle=self.phi.value / 2, mode=self.modes
```

Comment:
Why is phi divided by 2 here?

Reply:
It is multiplied by 2 when defining the `DisplacedSqueezed` state [here](https://github.com/XanaduAI/MrMustard/blob/3ea6f4b9c7d6399726c4d9013ac9396a006bbe84/mrmustard/lab/detectors.py#L273).